### ISSUE TITLE
Add an SBT plugin

### ISSUE THREAD

**Issue by InTheNow** [InTheNow/catalysts#25]
`project/Base.scala` contains some generic sbt functionality that could be used in other {typelevel,cats}-like projects, and is currently copied as is to [alleycats](https://github.com/non/alleycats/blob/master/project/Base.scala).

So we could easily produce an sbt-plugin that just includes this file, and this on its own would be useful.

That said, it's still very far off a _complete_ solution. But to add complete functionality would risk that easy adoption opportunities are lost... So my _suggestion_ is that we:

- Add new branch `wip/new-plugin` and create a plugin with `Base.scala`
- Use the package `org.typelevel.catalyts.sbt`
- Release this by 04.10.2015 to org.typelevel
- Any additions to `Base`/plugin are fine, provided they are finished by the due date
- Release is to `sbt-catalysts`

This is just a suggestion, not a doctrine. But if we discuss this and change as need, we can still get something useful out quite quickly

**Comment by InTheNow** [comment]
Some suggestions to move to base:

- Make [pomextra developers](https://github.com/InTheNow/catalysts/blob/master/build.sbt#L218) a [function](https://github.com/typelevel/sbt-typelevel/blob/master/src/main/scala/Developer.scala)
- Add a `sharedScoverageSettings`, with `coverageMinimum` as an argument taken from the build settings
- make most of the `docSettings` and`docs.settings` as shared
- consider adding the lib versions to Base, perhaps as a simple map. eg  `Map("disciplineVersion" ->"0.4")`. It's then trivial for the client to copy the map, overriding/adding as required.

**Comment by non** [comment]
I like the idea that upgrading versions will be relatively easy.

**Comment by dwijnand** [comment]
Why not just `sbtcatalysts`?

**Comment by InTheNow** [comment]
> sbtcatalysts

because it's likely that we may have `org.typelevel.catalalysts.somethingelse` and `org.typelevel.catalalysts.anotherthing`

**Comment by InTheNow** [comment]
> Why not just sbtcatalysts?

Then again, a lot of sbt plugins do this to simplify the import/namespace. Maybe it makes more sense

**Comment by dwijnand** [comment]
I've been moving to shorter package names. Discipline seems to be the only one that's using org.typelevel, not even catalysts uses it. I think it's better- short and sweet

**Comment by dwijnand** [comment]
In addition to `Base.scala` another reference point for "things should should do to setup your project" is @non's slides: http://plastic-idolatry.com/erik/sw2015.pdf (refers to tut, sbt-unidoc, sbt-site, sbt-ghpages, etc)

**Comment by InTheNow** [comment]
@dwijnand Thanks for the comments on the `wip/new-plugin` commit!

**Comment by InTheNow** [comment]
In case people can't find it  - https://github.com/InTheNow/catalysts/commit/51893fc407f6ee34f36826e0d62fc198a39fbe82#commitcomment-13503189 is interesting in general

**Comment by InTheNow** [comment]
@sjrd @gzm0 @nicolasstucki Whilst I concede that this is not "your" project, some of the ideas that I/we are proposing here may well end up being used by quite a few projects and I would not want to propagate poor standards - especially where scala.js is concerned (In particular, I have changed the syntax of `CrossProject` => `Module` to allow for an aggregate project)

I wonder if you have some spare moments to look at the build settings, not from a pure code review POV, but from a more general, architectural angle. If you can't, no  worries - I more than appreciate all your work on sjs:) 

FYI, whilst some of the ideas (esp Module) are reminiscent of my previous `sbt-scalajs` plugin, all the features are based on feedback from Cats, etc. 

The core functions are in [Base](https://github.com/InTheNow/catalysts/blob/wip/new-plugin/project/Base.scala), including the [module definitions](https://github.com/InTheNow/catalysts/blob/wip/new-plugin/project/Base.scala#L210), all used by [build.sbt](https://github.com/InTheNow/catalysts/blob/wip/new-plugin/build.sbt).

**Comment by sjrd** [comment]
You certainly don't want to have your module `dependsOn` its JS and JVM parts ([here](https://github.com/InTheNow/catalysts/blob/3d7d8d8dcd1df5c14be4b8ea16af9ae3a13ca679/project/Base.scala#L264)). `aggregate`, yes; `dependsOn`, no.

**Comment by sjrd** [comment]
What's the intended audience? Typelevel projects? Or broader? If broader, you don't want to enable language features by default ([here](https://github.com/InTheNow/catalysts/blob/3d7d8d8dcd1df5c14be4b8ea16af9ae3a13ca679/project/Base.scala#L114-L117)). I understand Typelevel projects use them all the time, and it can make sense to enable them by default. But most projects don't want that.

[Some of the other settings](https://github.com/InTheNow/catalysts/blob/3d7d8d8dcd1df5c14be4b8ea16af9ae3a13ca679/project/Base.scala#L119-L126) might also not be wanted by most projects.

**Comment by InTheNow** [comment]
Intended audience is (initially) typelevel projects and users of typelevel projects. The idea is that we supply a bunch of settings, and client projects exclude the ones they _don't_ want.

From your comments, it looks like that we could/should split them into three, eg `scalacCommonOptions`, `scalacLanguageOptions` and `scalacStrictOptions` and also have an aggregate `scalacAllOptions`.

Thx, btw :)

**Comment by dwijnand** [comment]
`scalacGoodOptions`, `scalacBetterOptions`, `scalacBestOptions` :D

**Comment by sjrd** [comment]
@dwijnand I disagree that `-language:` options are better. IMO they are worse than not putting them, since they are *less strict*.

**Comment by dwijnand** [comment]
I can see your point.

From my point of view enabling -language scalac options is better than the verbosity of importing the language feature per source file.
